collect/central_collector.go

```diff
@@ -229,13 +499,62 @@ func (c *CentralCollector) GetParentID(sp *types.Span) (string, bool) {
 	return "", false
 }
 
-func (c *CentralCollector) Stressed() bool {
-	return false
-}
+func (c *CentralCollector) send(status *centralstore.CentralTraceStatus) {
+	trace := c.SpanCache.Get(status.TraceID)
+	if trace == nil {
+		return
+	}
+	traceDur := time.Since(trace.ArrivalTime)
+	c.Metrics.Histogram("trace_duration_ms", float64(traceDur.Milliseconds()))
+	c.Metrics.Histogram("trace_span_count", float64(status.DescendantCount()))
+
+	c.Metrics.Increment(status.KeepReason)
+
+	// get sampler key (dataset for legacy keys, environment for new keys)
+	samplerKey, isLegacyKey := trace.GetSamplerKey()
+	logFields := logrus.Fields{
+		"trace_id": trace.TraceID,
+	}
+	if isLegacyKey {
+		logFields["dataset"] = samplerKey
+	} else {
+		logFields["environment"] = samplerKey
+	}
+
+	// If we have a root span, update it with the count before determining the SampleRate.
+	if trace.RootSpan != nil {
+		rs := trace.RootSpan
+		rs.Data["meta.span_event_count"] = int64(status.SpanEventCount())
+		rs.Data["meta.span_link_count"] = int64(status.SpanLinkCount())
+		rs.Data["meta.span_count"] = int64(status.SpanCount())
+		rs.Data["meta.event_count"] = int64(status.DescendantCount())
+	}
+
+	logFields["reason"] = status.KeepReason
 
-func (c *CentralCollector) GetStressedSampleRate(traceID string) (rate uint, keep bool, reason string) {
-	return 0, false, ""
+	c.Metrics.Increment("trace_send_kept")
+	// This will observe sample rate decisions only if the trace is kept
+	c.Metrics.Histogram("trace_kept_sample_rate", float64(status.Rate))
+
+	c.Logger.Info().WithFields(logFields).Logf("Sending trace")
+	for _, sp := range trace.GetSpans() {
+		if c.Config.GetAddRuleReasonToTrace() {
+			sp.Data["meta.refinery.reason"] = status.Metadata["meta.refinery.reason"]
```

Comment:
I think this should probably be a loop over Metadata so we can add new attributes without forgetting to update this.